File: crates/browser-core/src/browser_url_intelligence_youtube.rs
use super::*;

const SEGMENT_VIDEO: &str = "video";
const SEGMENT_SHORTS: &str = "shorts";
const SEGMENT_WATCH: &str = "watch";
const SEGMENT_EMBED: &str = "embed";
const SEGMENT_LIVE: &str = "live";
const SEGMENT_CHANNEL: &str = "channel";
const SEGMENT_CUSTOM: &str = "c";
const SEGMENT_USER: &str = "user";
const SEGMENT_HANDLE: &str = "@";
const SEGMENT_PLAYLIST: &str = "playlist";
const SEGMENT_RESULTS: &str = "results";
const QUERY_VIDEO: &str = "v";
const QUERY_PLAYLIST: &str = "list";
const QUERY_SEARCH: &str = "search_query";
const PLATFORM_VIMEO: &str = "vimeo";
const PLATFORM_YOUTUBE: &str = "youtube";
const PLATFORM_YOUTUBE_SHORTS: &str = "youtube-shorts";
const TARGET_UNKNOWN: &str = "unknown";
const TARGET_SHORT_VIDEO: &str = "short-video";
const TARGET_CHANNEL: &str = "channel";
const TARGET_PLAYLIST: &str = "playlist";
const TARGET_SEARCH: &str = "search";
const CONFIDENCE_HIGH: &str = "high";
const CONFIDENCE_LOW: &str = "low";
const REASON_PARSED_URL: &str = "parsed-url";
const REASON_YOUTUBE_SHORTS_ID: &str = "parsed-youtube-shorts-id";
const REASON_YOUTUBE_VIDEO_ID: &str = "parsed-youtube-video-id";
const REASON_PLAYLIST_ID: &str = "parsed-playlist-id";
const REASON_SEARCH_QUERY: &str = "parsed-search-query";
const REASON_CHANNEL_ID: &str = "parsed-channel-id";

#[path = "../../browser-core-generated/browser_url_intelligence_youtube_core_impl.rs"]
mod core;
#[path = "../../browser-core-generated/browser_url_intelligence_youtube_routes_impl.rs"]
mod routes;

pub(super) fn vimeo_shape(parsed: &ParsedBrowserUrl) -> Option<ParsedUrlShape> {
    core::vimeo_shape(parsed)
}

pub(super) fn youtube_shape(parsed: &ParsedBrowserUrl) -> ParsedUrlShape {
    let segments = path_segments(parsed);
    core::youtube_shorts_shape(&segments)
        .or_else(|| core::youtube_watch_shape(parsed, &segments))
        .or_else(|| core::youtube_embed_or_live_shape(&segments))
        .or_else(|| routes::youtube_named_channel_shape(&segments))
        .or_else(|| routes::youtube_handle_shape(&segments))
        .or_else(|| routes::youtube_playlist_shape(parsed, &segments))
        .or_else(|| routes::youtube_search_shape(parsed, &segments))
        .unwrap_or_else(|| {
            simple_shape(
                TARGET_UNKNOWN,
                PLATFORM_YOUTUBE,
                CONFIDENCE_LOW,
                vec![REASON_PARSED_URL],
            )
        })
}

File: crates/browser-core-generated/browser_url_intelligence_youtube.ts
import type { ParsedBrowserUrl, ParsedUrlShape } from './browser_url_intelligence';
import { hasText, pathSegments } from './browser_url_intelligence_url_text';
import { queryParam } from './browser_url_intelligence_query';
import { channelShape, emptyPlatformIds, simpleShape, videoShape } from './browser_url_intelligence_social_common';

export function vimeoShape(parsed: ParsedBrowserUrl): ParsedUrlShape | null {
  const segments = pathSegments(parsed);
  const videoId = segments[0] === 'video' ? (segments[1] ?? null) : (segments[0] ?? null);
  return videoId !== null && /^[0-9]+$/.test(videoId) ? videoShape('vimeo', videoId, ['parsed-url']) : null;
}

export function youtubeShape(parsed: ParsedBrowserUrl): ParsedUrlShape {
  const segments = pathSegments(parsed);
  return (
    youtubeShortsShape(segments) ??
    youtubeWatchShape(parsed, segments) ??
    youtubeEmbedOrLiveShape(segments) ??
    youtubeNamedChannelShape(segments) ??
    youtubeHandleShape(segments) ??
    youtubePlaylistShape(parsed, segments) ??
    youtubeSearchShape(parsed, segments) ??
    simpleShape('unknown', 'youtube', 'low', ['parsed-url'])
  );
}

function youtubeShortsShape(segments: readonly string[]): ParsedUrlShape | null {
  const videoId = segments[1];
  return segments[0] === 'shorts' && videoId !== undefined
    ? {
        ...videoShape('youtube-shorts', videoId, ['parsed-url', 'parsed-youtube-shorts-id']),
        targetKind: 'short-video',
      }
    : null;
}

function youtubeWatchShape(parsed: ParsedBrowserUrl, segments: readonly string[]): ParsedUrlShape | null {
  const videoId = queryParam(parsed, 'v');
  return (segments[0] === 'watch' || segments.length === 0) && hasText(videoId)
    ? videoShape('youtube', videoId, ['parsed-url', 'parsed-youtube-video-id'])
    : null;
}

function youtubeEmbedOrLiveShape(segments: readonly string[]): ParsedUrlShape | null {
  const videoId = segments[1];
  return (segments[0] === 'embed' || segments[0] === 'live') && videoId !== undefined
    ? videoShape('youtube', videoId, ['parsed-url', 'parsed-youtube-video-id'])
    : null;
}

function youtubeNamedChannelShape(segments: readonly string[]): ParsedUrlShape | null {
  const channelId = segments[1];
  return ['channel', 'c', 'user', '@'].includes(segments[0] ?? '') && channelId !== undefined
    ? youtubeChannelShape(channelId)
    : null;
}

function youtubeHandleShape(segments: readonly string[]): ParsedUrlShape | null {
  const channelId = segments[0];
  return channelId?.startsWith('@') === true ? youtubeChannelShape(channelId) : null;
}

function youtubePlaylistShape(parsed: ParsedBrowserUrl, segments: readonly string[]): ParsedUrlShape | null {
  const playlistId = queryParam(parsed, 'list');
  return segments[0] === 'playlist' && hasText(playlistId)
    ? {
        targetKind: 'playlist',
        platform: 'youtube',
        platformIds: { ...emptyPlatformIds(), playlistId },
        confidence: 'high',
        reasonCodes: ['parsed-url', 'parsed-playlist-id'],
      }
    : null;
}

function youtubeSearchShape(parsed: ParsedBrowserUrl, segments: readonly string[]): ParsedUrlShape | null {
  const query = queryParam(parsed, 'search_query');
  return segments[0] === 'results' && hasText(query)
    ? {
        targetKind: 'search',
        platform: 'youtube',
        platformIds: { ...emptyPlatformIds(), query },
        confidence: 'high',
        reasonCodes: ['parsed-url', 'parsed-search-query'],
      }
    : null;
}

function youtubeChannelShape(channelId: string): ParsedUrlShape {
  return channelShape('youtube', channelId, 'high');
}
